src/keys.rs

```diff
@@ -188,21 +188,122 @@ impl SpendValidatingKey {
     pub(crate) fn from_bytes(bytes: &[u8]) -> Option<Self> {
         <[u8; 32]>::try_from(bytes)
             .ok()
-            .and_then(|b| {
-                // Structural validity checks for ak_P:
-                // - The point must not be the identity
-                //   (which for Pallas is canonically encoded as all-zeroes).
-                // - The sign of the y-coordinate must be positive.
-                if b != [0; 32] && b[31] & 0x80 == 0 {
-                    <redpallas::VerificationKey<SpendAuth>>::try_from(b).ok()
-                } else {
-                    None
-                }
-            })
+            .and_then(check_structural_validity)
             .map(SpendValidatingKey)
     }
 }
 
+/// An issuer authorizing key, used to create issuer authorization signatures.
+/// This type enforces that the corresponding public point (ik^ℙ) has ỹ = 0.
+///
+/// $\mathsf{isk}$ as defined in
+/// [Zcash Protocol Spec § 4.2.3: Orchard Key Components][orchardkeycomponents].
+///
+/// [orchardkeycomponents]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
+#[derive(Clone, Debug)]
+pub struct IssuerAuthorizingKey(redpallas::SigningKey<SpendAuth>);
+
+impl IssuerAuthorizingKey {
+    /// Derives isk from sk. Internal use only, does not enforce all constraints.
+    fn derive_inner(sk: &SpendingKey) -> pallas::Scalar {
+        to_scalar(PrfExpand::ZsaIsk.expand(&sk.0))
+    }
+
+    /// Sign the provided message using the `IssuerAuthorizingKey`.
+    pub fn sign(
+        &self,
+        rng: &mut (impl RngCore + CryptoRng),
+        msg: &[u8],
+    ) -> redpallas::Signature<SpendAuth> {
+        self.0.sign(rng, msg)
+    }
+}
+
+impl From<&SpendingKey> for IssuerAuthorizingKey {
+    fn from(sk: &SpendingKey) -> Self {
+        let isk = Self::derive_inner(sk);
+        // IssuerSigningKey cannot be constructed such that this assertion would fail.
```

Comment:
```suggestion
        // IssuerAuthorizingKey cannot be constructed such that this assertion would fail.
```
(or `IssuanceAuthorizingKey`).

Reply:
done